---
title: "DE_Analysis_Pan_Wirka_Hu"
output: html_document
date: "2024-04-17"
editor_options: 
  chunk_output_type: console
---



# New Gene Set
```{r}
up_list_integ = readRDS(file = "/shared_home/my_analysis/ChipSeeker/BETA/up_list_df_Integrative.rds")
up_list_integ_human = gorth(up_list_integ$SYMBOL, source_organism = "mmusculus", target_organism ="hsapiens" )$ortholog_name
up_list_integ_human = na.omit(up_list_integ_human)
up_list_integ_human = up_list_integ_human[!duplicated(up_list_integ_human)]

```




# DESeq2 method:Condition-specific

## MQ
*Differentially Expressed Genes in Macrophage Population based on the Condition*

-   The goal is to assess the differential gene expression **in various experimental groups(lesion and non-lesion)**, restricted to **Macrophage Population**.

-   To find differentially expressed genes, DESeq2 method were implemented on [ZINB-adjusted expression data](https://www.nature.com/articles/s41467-017-02554-5), filtered based on padj < 0.05 & abs(log2FoldChange) > 0.25.

```{r echo=FALSE}

# Macrophage: DESeq2 Analysis using ZINB-WaVE

library(BiocParallel)
library(DESeq2)

BiocParallel::register(SerialParam())

rpca_int_sct_v3 = read_rds("pan_wirka_hu2/Integration_Objs/rpca_int_sct_clustered_TS_Ucel_Sctype_anotd_Final.rds")

DefaultAssay(rpca_int_sct_v3) = "SCT"

MQ_clusters = c("cDC")
MQ_clusters = c("Foamy_Mac","Inflammatory_Mac", "Tissue_resident_Mac", "Monocytes", "cDC")
mq_subset = subset(rpca_int_sct_v3, idents = MQ_clusters)

#https://r-charts.com/ggplot2/facets/
DimPlot(mq_subset, label = F, repel = F, split.by = "sample_disease_status") +
    custom_theme +
  NoAxes() + theme(strip.text = element_text(size = 14, face = "bold"), plot.subtitle = element_blank(), plot.title = element_text(size=14, face = "bold", hjust = 0.05, margin=margin(0,0,-40,0))) +
  labs(title = "Macrophage Subpopulation changes in different conditions ") 

# Re factor covariates
mq_subset$sample_disease_status = factor(mq_subset$sample_disease_status,
                                                  levels = c("non_lesion", "lesion"))

DimPlot(mq_subset, label = T, repel = T, split.by = "sample_disease_status") + NoLegend() + NoAxes()

mq_subset$sex = factor(mq_subset$sex, levels = c("males", "females"))
mq_subset$arterial_origin = factor(mq_subset$arterial_origin, levels = c("carotid", "coronary"))

ggsave(filename = "ord_et_al_GSE205930/Figures/CellNo_correction/MQ_with_disease.pdf", 
        width = 8.5, height = 9)
###################################
DefaultAssay(sctype_test) = "SCT"
Idents(sctype_test) = "customclassif"

#ISG expressing immune cells:: IFIT1,IFIT2,IFIT3,IFIT5,ISG15,CCL3,CCL4,CCL3L3,RSAD2,OASL,CXCL10,IFI15,ISG20
#Non-classical monocytes:: CD14,CD16,CD11b,CD68,HLA-DR,CD33,CD11c,CD123,CD15,CD3D,CD3E,CD3G,CD3Z,CD66b,FCGR3A,CDKN1C,LST1,FCER1G,MS4A7,RHOC,S100A8,S100A9,CST3,C1QC
mq_subset = subset(sctype_test, idents = c("ISG expressing immune cells", "Non-classical monocytes"))
DimPlot(mq_subset, label = T, group.by = "customclassif", repel = T, split.by = "disease_status") & NoLegend() & NoAxes()
ggsave(filename = "ord_et_al_GSE205930/Figures/CellNo_correction/MQ_Subs_with_disease.pdf", 
        width = 8.5, height = 9)

mq_subset = subset(sctype_test, idents = c("ISG expressing immune cells"))
DimPlot(mq_subset, label = T, group.by = "customclassif", repel = T, split.by = "disease_status") & NoLegend() & NoAxes()
###################################

# Find variable features
DefaultAssay(mq_subset) = "RNA"
mq_subset = NormalizeData(mq_subset,
                                   normalization.method = "LogNormalize",
                                   scale.factor = 10000)
mq_subset = FindVariableFeatures(mq_subset, 
                                          selection.method = "vst",
                                          nfeatures = 3000, assay = "RNA")

mq_subset_var_features = mq_subset@assays$RNA@var.features

# Make SingleCellExperiment object
mq_subset_se = Seurat::as.SingleCellExperiment(mq_subset)

class(mq_subset_se)

mq_subset_se_var_feats = mq_subset_se[mq_subset_var_features,]


# Model zero component using ZINB-WaVE

# Filter out lowly expressed genes by removing genes that
# do not have at least 5 reads in at least 5 samples
keep = rowSums(assay(mq_subset_se_var_feats) >= 5) >= 5

table(keep)
mq_subset_se_filtered = mq_subset_se_var_feats[keep,]

# Make sure that input counts matrix is of class matrix
assay(mq_subset_se_filtered) = round(as.matrix(assay(mq_subset_se_filtered)))
mq_subset_se_filtered

mq_subset_zinb = zinbwave(mq_subset_se_filtered, K=0, 
                           observationalWeights=TRUE,
                           X = "~ sex + arterial_origin + sample_disease_status",
                           BPPARAM=SerialParam())

saveRDS(mq_subset_zinb ,file = "pan_wirka_hu2/DE_Analysis_Objs/mq_subset_cDC_zinb.rds")

table(mq_subset_zinb@colData$sample_disease_status)

# Run DE analysis
dds = DESeq2::DESeqDataSet(mq_subset_zinb,
                   design = ~ sex + arterial_origin + sample_disease_status)

# Re factor lesion categories
dds@colData$sample_disease_status = factor(dds@colData$sample_disease_status, 
                                           levels = c("non_lesion", "lesion"))

# Estimate size factors
dds = estimateSizeFactors(dds, type="poscounts")

# Need to set the right parameters for single cell analysis
dds = DESeq(dds, test="LRT", minmu = 1e-6, 
            useT=FALSE,
            reduced=~1, 
            minReplicatesForReplace = Inf)

saveRDS(dds, file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_cDC.rds")

```

### Lesion vs Non-lesion

```{r echo=FALSE, cache=TRUE}
# whole MQ
dds_hu = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq.rds")

# Foamy_Mac
dds_hu = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_Foamy_Mac.rds")

# Inflammatory_Mac
dds_hu = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_Inflammatory_Mac.rds")

# Tissue_resident_Mac
dds_hu = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_Tissue_resident_Mac.rds")

# Monocytes
dds_hu = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_Monocytes.rds")

# cDC
dds_hu = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_cDC.rds")

##########LD vs Ct
res_mq = DESeq2::results(dds_hu, contrast = c("sample_disease_status",
                                        "lesion", "non_lesion"))

res_mq = as.data.frame(res_mq)
res_mq$gene = rownames(res_mq)

# Set threshold for filtering (FDR < 0.05 and log2FC > 0.25)
res_mq_filtd = res_mq %>%
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) >= 0.25 & baseMean > 0.3 ) %>%
  arrange(desc(log2FoldChange))

dim(res_mq_filtd)

#write_tsv(res_mq_filtd, file = "Human_SingleCell_MQ_DE.tsv", quote = "none")

######################
##Volcano
res_mq_filtd %>%
  dplyr::filter(abs(log2FoldChange) < 5) %>%
  ggplot(aes(x=log2FoldChange, y=-log10(padj), label= gene)) + 
  geom_point(size = 0.7) + 
  geom_vline(xintercept = c(-0.25, 0, 0.25), color="darkred", linetype="dashed") + 
  geom_hline(yintercept = -log10(1e-10), color="darkred", linetype="dashed") + 
  geom_text_repel(data = . %>% filter(gene %in% up_list_integ_human)) +
  theme_classic()

#Foamy macrophage
res_mq_filtd %>% 
  dplyr::filter(abs(log2FoldChange) < 5) %>%
  ggplot(aes(x=log2FoldChange, y=-log10(padj), label= gene, color = ifelse(log2FoldChange > 0.25, "#4575b4", "#d73027"))) + 
  geom_point(size = 0.3, alpha = 0.2) +
  #geom_quasirandom(groupOnX=FALSE, size = 0.5) + #From ggbeeswarm package
  geom_vline(xintercept = c(-0.25, 0.25), color="black", linetype="dashed") + 
  geom_hline(yintercept = -log10(5e-2), color="black", linetype="dashed") + 
  geom_text_repel(data = . %>% filter(gene %in% up_list_integ_human), size = 2.3,
                 box.padding = 0.09, force = 1, force_pull = 0, max.overlaps = Inf,             min.segment.length = Inf, max.iter = 1e4) +
  theme_classic() + theme(legend.position = "none", title = element_text(size = 12),
                          axis.text = element_text(size = 10)) +
   labs(title = "Foamy macrophage")


ggsave(filename = "pan_wirka_hu2/Figures/volcano_hMQ_Foamy_res.png", 
        width = 6.4, height = 4)

#Inflammatory macrophage
res_mq_filtd %>% 
  dplyr::filter(abs(log2FoldChange) < 5) %>%
  ggplot(aes(x=log2FoldChange, y=-log10(padj), label= gene, color = ifelse(log2FoldChange > 0.25, "#4575b4", "#d73027"))) + 
  geom_point(size = 0.3, alpha = 0.2) +
  #geom_quasirandom(groupOnX=FALSE, size = 0.5) + #From ggbeeswarm package
  geom_vline(xintercept = c(-0.25, 0.25), color="black", linetype="dashed") + 
  geom_hline(yintercept = -log10(5e-2), color="black", linetype="dashed") + 
  geom_text_repel(data = . %>% filter(gene %in% up_list_integ_human), size = 2,
                 box.padding = 0.09, force = 0.3, force_pull = 0, max.overlaps = Inf,             min.segment.length = Inf, max.iter = 1e4) +
  theme_classic() + theme(legend.position = "none", title = element_text(size = 12),
                          axis.text = element_text(size = 10)) +
   labs(title = "Inflammatory macrophage")

ggsave(filename = "pan_wirka_hu2/Figures/volcano_hMQ_inflammatory_res.png", 
        width = 6.4, height = 4)

#Tissue-resident macrophage 
 res_mq_filtd %>% 
  dplyr::filter(abs(log2FoldChange) < 5) %>%
  ggplot(aes(x=log2FoldChange, y=-log10(padj), label= gene, color = ifelse(log2FoldChange > 0.25, "#4575b4", "#d73027"))) + 
  geom_point(size = 0.3, alpha = 0.2) +
  #geom_quasirandom(groupOnX=FALSE, size = 0.5) + #From ggbeeswarm package
  geom_vline(xintercept = c(-0.25, 0.25), color="black", linetype="dashed") + 
  geom_hline(yintercept = -log10(5e-2), color="black", linetype="dashed") + 
  geom_text_repel(data = . %>% filter(gene %in% up_list_integ_human), size = 2.3,
                 box.padding = 0.09, force = 0.04, force_pull = 0, max.overlaps = Inf,             min.segment.length = Inf, max.iter = 1e4) +
  theme_classic() + theme(legend.position = "none", title = element_text(size = 12),
                          axis.text = element_text(size = 10)) +
   labs(title = "Tissue-resident macrophage")

ggsave(filename = "pan_wirka_hu2/Figures/volcano_hMQ_tissue_resident_res.png", 
        width = 6.4, height = 4)

# Monocytes 
res_mq_filtd %>% 
  dplyr::filter(abs(log2FoldChange) < 5) %>%
  ggplot(aes(x=log2FoldChange, y=-log10(padj), label= gene, color = ifelse(log2FoldChange > 0.25, "#4575b4", "#d73027"))) + 
  geom_point(size = 0.3, alpha = 0.2) +
  #geom_quasirandom(groupOnX=FALSE, size = 0.5) + #From ggbeeswarm package
  geom_vline(xintercept = c(-0.25, 0.25), color="black", linetype="dashed") + 
  geom_hline(yintercept = -log10(5e-2), color="black", linetype="dashed") + 
  geom_text_repel(data = . %>% filter(gene %in% up_list_integ_human), size = 2.3,
                 box.padding = 0.09, force = 0.4, force_pull = 0, max.overlaps = Inf,             min.segment.length = Inf, max.iter = 1e4) +
  theme_classic() + theme(legend.position = "none", title = element_text(size = 12),
                          axis.text = element_text(size = 10)) +
   labs(title = "Monocytes")

ggsave(filename = "pan_wirka_hu2/Figures/volcano_hMQ_monocytes_res.png", 
        width = 6.4, height = 4)

#cDC
res_mq_filtd %>% 
  dplyr::filter(abs(log2FoldChange) < 5) %>%
  ggplot(aes(x=log2FoldChange, y=-log10(padj), label= gene, color = ifelse(log2FoldChange > 0.25, "#4575b4", "#d73027"))) + 
  geom_point(size = 0.3, alpha = 0.2) +
  #geom_quasirandom(groupOnX=FALSE, size = 0.5) + #From ggbeeswarm package
  geom_vline(xintercept = c(-0.25, 0.25), color="black", linetype="dashed") + 
  geom_hline(yintercept = -log10(5e-2), color="black", linetype="dashed") + 
  geom_text_repel(data = . %>% filter(gene %in% up_list_integ_human), size = 2,
                 box.padding = 0.09, force = 0.4, force_pull = 0, max.overlaps = Inf,              min.segment.length = Inf, max.iter = 1e4) +
  theme_classic() + theme(legend.position = "none", title = element_text(size = 12),
                          axis.text = element_text(size = 10)) +
   labs(title = "cDC")

ggsave(filename = "pan_wirka_hu2/Figures/volcano_hMQ_cDC_res.png", 
        width = 6.4, height = 4)
##########

ids_mq_filtd = data.frame(symbol = res_mq_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene))
ids_mq_filtd$ENTREZID = mapIds(org.Hs.eg.db, keys = res_mq_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene), keytype = "SYMBOL", column = "ENTREZID", multiVals = "first")
mq_filtd_kegg = enrichKEGG(ids_mq_filtd$ENTREZID, organism = "hsa")

par(mfrow = c(1, 1), mar = c(6, 20, 2, 1))
barplot(height = -log10(mq_filtd_kegg@result$p.adjust)[20:1], names.arg = mq_filtd_kegg@result$Description[20:1],
    horiz = TRUE, las = 1, border = FALSE, cex.names = 0.65, xlab = "-log10(P.adjust)", main = "KEGG Analysis")
abline(v = c(-log10(0.05)), lty = 2)
abline(v = 0, lty = 1)


mq_subset = readRDS(file = "ord_et_al_GSE205930/DE_Cell_Specific/mq_subset_normalized.rds")
mq_subset = ScaleData(mq_subset, assay = "RNA")
DoHeatmap(mq_subset, features = res_mq_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene),  assay = "RNA", size = 4, angle = 90,  group.by = "sample_disease_status") 

DotPlot(mq_subset, features = res_mq_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene),  assay = "RNA", group.by = "sample") + RotatedAxis() + ggtitle("The expression pattern of top 10 genes from pairwaise comparisons") + theme(plot.title = element_text(face = "bold", size = 14))


```



#### Heatmap
```{r}
# https://jokergoo.github.io/ComplexHeatmap-reference/book/more-examples.html
# https://stackoverflow.com/questions/76376652/complexheatmap-maximum-height-of-row-labels-anno-mark
# https://jokergoo.github.io/ComplexHeatmap-reference/book/other-tricks.html #Saving heatmaps

library(Seurat)
library(tibble)
library(rlang)
library(circlize)
library(ggplot2)
library(ComplexHeatmap)
library(RColorBrewer)
library(grid)
library(grDevices)

hm_limit = c(-2, 0, 2) 
hm_colors = c("#4575b4","white","#d73027")
row_font_size = 6

### Select Genes

highlightd_genes = up_list_integ_human

genes = res_mq_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene)
###
mq_subset = ScaleData(mq_subset, assay = "RNA")
mat <- GetAssayData(object = mq_subset, assay = "RNA", slot = "scale.data")
mat <- mat[match(genes, rownames(mat)),]

anno <- mq_subset@meta.data %>%
    rownames_to_column(var = "barcode") %>%
    arrange(!!!syms(c("sample_disease_status", "annotations")))

mat <- t(mat)
mat <- mat[match(anno$barcode, rownames(mat)),]
mat <- t(mat)

annos <- list()

### sample_disease_status
value <- anno[["sample_disease_status"]]
l <- levels(factor(anno[["sample_disease_status"]]))
col_a = c("green", "darkred")
names(col_a) <- l
col_a <- col_a[!is.na(names(col_a))]

ha <- HeatmapAnnotation(a = anno[["sample_disease_status"]],
                              col = list(a = col_a),border = TRUE,
                              annotation_label = "Condition",
                        annotation_legend_param = list(title = "Condition"))

names(ha) <- "sample_disease_status"
annos[["sample_disease_status"]] <- ha
####
####annotations
value <- anno[["annotations"]]
l <- levels(factor(anno[["annotations"]]))
col_b = set_colors("Set2", length(l))
names(col_b) <- l
col_b <- col_b[!is.na(names(col_b))]


ha1 <- HeatmapAnnotation(b = anno[["annotations"]],
                              col = list(b = col_b), border = TRUE,
                              annotation_label = "Cell Type",
                         annotation_legend_param = list(title = "Cell Type"))

names(ha1) <- "annotations"
annos[["annotations"]] <- ha1
###
### arterial_origin
value <- anno[["arterial_origin"]]
l <- levels(factor(anno[["arterial_origin"]]))
col_c = set_colors("Reds", length(l))
names(col_c) <- l
col_c <- col_c[!is.na(names(col_c))]

ha2 <- HeatmapAnnotation(c = anno[["arterial_origin"]],
                              col = list(c = col_c), border = TRUE,
                              annotation_label = "Arterial Bed",
                          annotation_legend_param = list(title = "Arterial Bed"))

names(ha2) <- "arterial_origin"
annos[["arterial_origin"]] <- ha2
####

annos <- do.call(c, annos)
annos@gap <- rep(unit(1,"mm"), length(annos))

set.seed(20)

ranot = rowAnnotation(link = anno_mark(at = which(rownames(mat) %in% highlightd_genes), 
                     labels = highlightd_genes, side = "left", 
                     labels_gp = gpar(fontsize = 4.2), padding = unit(0.7, "mm"))
                     ) 
####First approach
set.seed(20) # To be reproducible

ht = Heatmap(mat, cluster_rows = TRUE,
                cluster_columns = FALSE,
                heatmap_legend_param = list(direction = "horizontal",
                                            legend_width = unit(6, "cm"),
                                            title = "Expression"),
                col = colorRamp2(hm_limit, hm_colors),
                show_column_names = FALSE, row_names_side = "left",
                row_names_gp = gpar(fontsize = row_font_size),
                top_annotation = annos, show_row_names = F, show_row_dend = T, row_km = 3,
                row_km_repeats = 20, left_annotation = ranot)
################################
####2nd approach
library(dendsort)

set.seed(20)
row_dend = dendsort(hclust(dist(mat),method = "ward.D")) #"ward.D" looks also OK!
plot(row_dend)

ht = Heatmap(mat, cluster_rows = row_dend, split = 3,
                cluster_columns = FALSE,
                heatmap_legend_param = list(direction = "horizontal",
                                            legend_width = unit(6, "cm"),
                                            title = "Expression"),
                col = colorRamp2(hm_limit, hm_colors),
                show_column_names = FALSE, row_names_side = "left",
                row_names_gp = gpar(fontsize = row_font_size),
                top_annotation = annos, show_row_names = F, show_row_dend = T,
                 left_annotation = ranot)
########       
pdf(file = "pan_wirka_hu2/Figures/htmap_mq_NA.pdf", width = 7.5, height = 9) #using 1st strategy

ht = draw(ht, heatmap_legend_side = "bottom",annotation_legend_side = "right")

dev.off()

saveRDS(ht, file = "pan_wirka_hu2/Figures/htmap_mq_rk3_integ.rds")
```

##### Venn
```{r}
# From: https://r-graph-gallery.com/14-venn-diagramm
#       https://stackoverflow.com/questions/14243609/venn-diagram-does-not-produce-any-output
library(VennDiagram)

r.dend <- row_dend(ht)  #If needed, extract row dendrogram
rcl.list <- row_order(ht)

# Prepare data sets
cl1 = rownames(mat[rcl.list[["1"]],])
cl1_st = cl1[cl1 %in% up_list_integ_human]

cl2 = rownames(mat[rcl.list[["2"]],])
cl2_st = cl2[cl2 %in% up_list_integ_human]

cl3 = rownames(mat[rcl.list[["3"]],])
cl3_st = cl3[cl3 %in% up_list_integ_human]


# Prepare a palette of 3 colors with R colorbrewer:
library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")[c(1,2)]

# Chart
ven_cl1 = venn.diagram(x = list(cl1, up_list_integ_human),
        category.names = c("Genes in Cluster 1" , "Integrative Genes"),
        filename = NULL, disable.logging = T, output =T,
        # Output features
        imagetype="png" , height = 300, width = 300, resolution = 300, compression = "lzw",
        # Circles
        lwd = 2, lty = 'blank', fill = myCol,
        # Numbers (fontface = "bold")
        cex = 0.9, fontfamily = "sans",
        # Set names (cat.fontface = "bold")
        cat.cex = .7, cat.default.pos = "outer", cat.pos = c(-170, 27),
        cat.fontfamily = "sans")

ven_cl2 = venn.diagram(x = list(cl2, up_list_integ_human),
        category.names = c("Genes in Cluster 2" , "Integrative Genes"),
        filename = NULL, disable.logging = T, output =T,
        # Output features
        imagetype="png" , height = 300, width = 300, resolution = 300, compression = "lzw",
        # Circles
        lwd = 2, lty = 'blank', fill = myCol,
        # Numbers (fontface = "bold")
        cex = 0.9, fontfamily = "sans",
        # Set names (cat.fontface = "bold")
        cat.cex = .7, cat.default.pos = "outer", cat.pos = c(-170, 27),
        cat.fontfamily = "sans")

ven_cl3 = venn.diagram(x = list(cl3, up_list_integ_human),
        category.names = c("Genes in Cluster 3" , "Integrative Genes"),
        filename = NULL, disable.logging = T, output =T,
        # Output features
        imagetype="png" , height = 300, width = 300, resolution = 300, compression = "lzw",
        # Circles
        lwd = 2, lty = 'blank', fill = myCol,
        # Numbers
        cex = 0.9, fontfamily = "sans",
        # Set names
        cat.cex = .7, cat.default.pos = "outer", cat.pos = c(-170, 27),
        cat.fontfamily = "sans")

vens = plot_grid(ven_cl1, ven_cl2, ven_cl3 , nrow = 1, scale =  0.80, align = "hv")

pdf(file = "pan_wirka_hu2/Figures/Ven_mq_rk3.pdf", width = 6, height = 2.5) 

grid.draw(vens)

dev.off()

```

##### ClusterProfiler
```{r}
# cl1
cl1_ranks_df = res_mq_filtd %>% filter(gene %in% cl1) %>% dplyr::select(c(gene, log2FoldChange))
cl1_ranks = cl1_ranks_df$log2FoldChange
names(cl1_ranks) = cl1_ranks_df$gene

# cl2
cl2_ranks_df = res_mq_filtd %>% filter(gene %in% cl2) %>% dplyr::select(c(gene, log2FoldChange))
cl2_ranks = cl2_ranks_df$log2FoldChange
names(cl2_ranks) = cl2_ranks_df$gene

# cl3
cl3_ranks_df = res_mq_filtd %>% filter(gene %in% cl3) %>% dplyr::select(c(gene, log2FoldChange))
cl3_ranks = cl3_ranks_df$log2FoldChange
names(cl3_ranks) = cl3_ranks_df$gene

# Run gene set enrichment
mac_gost_res = gprofiler2::gost(names(cl2_ranks), organism = "hsapiens",
                                         ordered_query = T, correction_method = "fdr")
mac_gprofiler_res = mac_gost_res$result
mac_go_bp = mac_gprofiler_res[mac_gprofiler_res$source == "GO:BP", ]

# Plot GO BP terms
mac_go_plot = gprofiler_bar_plot(mac_go_bp, 15, 500, "GO:BP", 10) + new_scale3_bars


ids_mq_filtd = data.frame(symbol = cl2_st)
ids_mq_filtd$ENTREZID = mapIds(org.Hs.eg.db, keys = cl2_st, keytype = "SYMBOL", column = "ENTREZID", multiVals = "first")
#mq_filtd_kegg = enrichGO(ids_mq_filtd$ENTREZID, OrgDb = "org.Hs.eg.db", ont = "BP")
mq_filtd_kegg = enrichKEGG(ids_mq_filtd$ENTREZID, organism = "hsa")
par(mfrow = c(1, 1), mar = c(6, 20, 2, 1))
barplot(height = -log10(mq_filtd_kegg@result$p.adjust)[20:1], names.arg = mq_filtd_kegg@result$Description[20:1],
    horiz = TRUE, las = 1, border = FALSE, cex.names = 0.65, xlab = "-log10(P.adjust)", main = "KEGG Analysis")
abline(v = c(-log10(0.05)), lty = 2)
abline(v = 0, lty = 1)

#### ClusterProfiler
# Conversion
cluster_list <- list(Cluster1 = cl1, Cluster2 = cl2, Cluster3 = cl3)
#cluster_list <- list(cl1, cl2, cl2_st, cl3, cl3_st)

# Function to convert SYMBOL to ENTREZID
convert_to_entrez <- function(gene_cluster) {
  mapIds(org.Hs.eg.db, keys = gene_cluster, keytype = "SYMBOL", column = "ENTREZID", multiVals = "first")
}

# Apply the function to each gene cluster in the list
cluster_list_entrez <- map(cluster_list, convert_to_entrez)

#KEGG
htmp_cluster_list_enrichKEGG = compareCluster(geneCluster = cluster_list_entrez, fun = "enrichKEGG", pvalueCutoff = 0.05, organism = "hsa")
#BP
htmp_cluster_list_enrichGO = compareCluster(geneCluster = cluster_list_entrez, fun = "enrichGO", pvalueCutoff = 0.05, ont="BP", OrgDb = "org.Hs.eg.db")

mq_cls_kegg = dotplot(htmp_cluster_list_enrichKEGG, title = "Comparing KEGG results of heatmap clusters", showCategory=5, font.size =11) + scale_x_discrete(guide = guide_axis(angle = 0)) + theme(axis.title.x = element_blank(), axis.text.x = element_text(size = 10))

pdf(file = "pan_wirka_hu2/Figures/KEGG_Ven_mq_rk3.pdf", width = 8.5, height = 6) 

mq_cls_kegg / vens + plot_layout(heights = c(2,1.2))

dev.off()

```


#### Lesion_BMD-MQ
```{r}
# Foamy_Mac
dds_foamy = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_Foamy_Mac.rds")

# Inflammatory_Mac
dds_inflm = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_Inflammatory_Mac.rds")

# Tissue_resident_Mac
dds_TRMQ = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_Tissue_resident_Mac.rds")

# Monocytes
dds_mono = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_Monocytes.rds")

# cDC
dds_cDC = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_mq_cDC.rds")

##########LD vs Ct
res_mq = DESeq2::results(dds_TRMQ, contrast = c("sample_disease_status",
                                        "lesion", "non_lesion"))

res_mq = as.data.frame(res_mq)
res_mq$gene = rownames(res_mq)

# Set threshold for filtering (FDR < 0.05 and log2FC > 0.25)
res_mq_filtd = res_mq %>%
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) >= 0.25 & baseMean > 0.3 ) %>%
  arrange(desc(log2FoldChange))

dim(res_mq_filtd)

TRMQ_filtd =  res_mq_filtd

# Create a filtered list
Mq_list = list(Foamy_Mq = foamy_filtd, Inflmmatory_Mq = inflm_filtd, cDC = cdc_filtd,
               Monocyte = mono_filtd, Tissue_resident_Mq = TRMQ_filtd) 

Mq_list_df = rbindlist(Mq_list, idcol = T)

#write_tsv(Mq_list_df, file = "Hans_project/SingleCellHuman_Mq_Full_DEG_filtd.tsv")

# Check size
lapply(Mq_list, function(x) dim(x))

# load genes
ovlp_242 = read_csv(file = "Hans_project/Human_Mouse_overlap_242.csv", col_names = "gene", trim_ws = T)

# Use map() to apply inner_join to each dataframe in the list
overlap_list <- map(Mq_list, ~inner_join(.x, ovlp_242, by = "gene"))

map(overlap_list, ~dim(.x))

overlap_list_df = rbindlist(overlap_list, idcol = T)

#write_tsv(overlap_list_df, file = "Hans_project/SingleCellHuman_MouseOverlap.tsv")

```

#### Subtype GO annotation
```{r}
Mq_list_df = read_tsv(file = "Hans_project/SingleCellHuman_Mq_Full_DEG_filtd.tsv")

mq_subtypes = split(Mq_list_df, Mq_list_df$.id)

map(mq_subtypes, ~head(.))

# Define a function to apply mapIds to a dataframe
mapIds_on_df <- function(df) {
 df$ID = mapIds(org.Hs.eg.db, keys = df$gene, keytype = "SYMBOL", column = "ENTREZID", multiVals = "first")
 #return(df)
}

# Use map to apply the function to each dataframe in the list
mapped_mq_subtypes <- map(mq_subtypes, mapIds_on_df)

map(mapped_mq_subtypes, ~head(.,2))

# KEGG
mq_subtypes_KEGG = compareCluster(geneCluster = mapped_mq_subtypes,
                              fun = "enrichKEGG", pvalueCutoff = 0.05, organism = "hsa")

dotplot(mq_subtypes_KEGG, title = "Comparing KEGG results of human MQ subtypes", font.size =10)

ggsave(filename = "pan_wirka_hu2/Figures/Mq_subtypes_kegg.png", 
        width = 9, height = 6)

# BP
mq_subtypes_BP = compareCluster(geneCluster = mapped_mq_subtypes, fun = "enrichGO", pvalueCutoff = 0.05, ont="BP", OrgDb = "org.Hs.eg.db")

dotplot(mq_subtypes_BP, title = "Comparing KEGG results of human MQ subtypes", showCategory=1)

# BP
ranks_mq = function(df){
  df_ranks = df %>% dplyr::select(c(gene, log2FoldChange))
  df_ranks_fc = df_ranks$log2FoldChange
  names(df_ranks_fc) = df_ranks$gene
  return(df_ranks_fc)
}

ranks_mq_df = map(mq_subtypes, ranks_mq)
cl3_ranks_df = res_mq_filtd %>% filter(gene %in% cl3) %>% dplyr::select(c(gene, log2FoldChange))
cl3_ranks = cl3_ranks_df$log2FoldChange
names(cl3_ranks) = cl3_ranks_df$gene

# Run gene set enrichment
mac_gost_res = gprofiler2::gost(names(ranks_mq_df$Inflmmatory_Mq), organism = "hsapiens",
                                         ordered_query = T, correction_method = "fdr")
mac_gprofiler_res = mac_gost_res$result
mac_go_bp = mac_gprofiler_res[mac_gprofiler_res$source == "GO:BP", ]

# Plot GO BP terms
mac_go_plot = gprofiler_bar_plot(mac_go_bp, 15, 500, "GO:BP", 10) + new_scale3_bars

```


## SM

```{r}
# SM: DESeq2 Analysis using ZINB-WaVE

library(BiocParallel)
library(DESeq2)

BiocParallel::register(SerialParam())

rpca_int_sct_v3 = read_rds("pan_wirka_hu2/Integration_Objs/rpca_int_sct_clustered_TS_Ucel_Sctype_anotd_Final.rds")

DefaultAssay(rpca_int_sct_v3) = "SCT"

sm_subset = subset(rpca_int_sct_v3, idents = c("SMC"))

DimPlot(sm_subset, label = T, repel = T, split.by = "sample_disease_status") + NoLegend() + NoAxes()

# Re factor covariates
sm_subset$sample_disease_status = factor(sm_subset$sample_disease_status,
                                                  levels = c("non_lesion", "lesion"))

DimPlot(sm_subset, label = F, repel = F, split.by = "sample_disease_status") +
    custom_theme +
  NoAxes() + theme(strip.text = element_text(size = 12, face = "bold"), plot.subtitle = element_blank(), plot.title = element_text(size=12, face = "bold", hjust = 0.05, margin=margin(0,0,-40,0)), legend.text = element_text(size = 11)) +
  labs(title = "Smooth muscle cell changes in different conditions ") 

ggsave(filename = "pan_wirka_hu2/Figures/homo_SMC_condition_umap.png", width = 5.8, height = 4.3 )

sm_subset$sex = factor(sm_subset$sex, levels = c("males", "females"))
sm_subset$arterial_origin = factor(sm_subset$arterial_origin, levels = c("carotid", "coronary"))

ggsave(filename = "ord_et_al_GSE205930/Figures/CellNo_correction/MQ_with_disease.pdf", 
        width = 8.5, height = 9)


# Find variable features
DefaultAssay(sm_subset) = "RNA"
sm_subset = NormalizeData(sm_subset,
                                   normalization.method = "LogNormalize",
                                   scale.factor = 10000)
sm_subset = FindVariableFeatures(sm_subset, 
                                          selection.method = "vst",
                                          nfeatures = 3000, assay = "RNA")

sm_subset_var_features = sm_subset@assays$RNA@var.features

# Make SingleCellExperiment object
sm_subset_se = Seurat::as.SingleCellExperiment(sm_subset)

class(sm_subset_se)

sm_subset_se_var_feats = sm_subset_se[sm_subset_var_features,]


# Model zero component using ZINB-WaVE

# Filter out lowly expressed genes by removing genes that
# do not have at least 5 reads in at least 5 samples
keep = rowSums(assay(sm_subset_se_var_feats) >= 5) >= 5

table(keep)
sm_subset_se_filtered = sm_subset_se_var_feats[keep,]


# Make sure that input counts matrix is of class matrix
assay(sm_subset_se_filtered) = round(as.matrix(assay(sm_subset_se_filtered)))
sm_subset_se_filtered

sm_subset_se_filtered = sm_subset_se_filtered[,-c(9523)]
sm_subset_se_filtered = sm_subset_se_filtered[,-c(10599)]
sm_subset_se_filtered

sm_subset_zinb = zinbwave(sm_subset_se_filtered, K=0, 
                           observationalWeights=TRUE,
                           X = "~ sex + arterial_origin + sample_disease_status",
                           BPPARAM=SerialParam())

saveRDS(sm_subset_zinb ,file = "pan_wirka_hu2/DE_Analysis_Objs/sm_subset_zinb.rds")

table(mq_subset_zinb@colData$sample_disease_status)

# Run DE analysis
dds_sm = DESeq2::DESeqDataSet(sm_subset_zinb,
                   design = ~ sex + arterial_origin + sample_disease_status)

# Re factor lesion categories
dds_sm@colData$sample_disease_status = factor(dds_sm@colData$sample_disease_status, 
                                           levels = c("non_lesion", "lesion"))

# Estimate size factors
dds_sm = estimateSizeFactors(dds_sm, type="poscounts")

# Need to set the right parameters for single cell analysis
dds_sm = DESeq(dds_sm, test="LRT", minmu = 1e-6, 
            useT=FALSE,
            reduced=~1, 
            minReplicatesForReplace = Inf)

saveRDS(dds_sm, file = "pan_wirka_hu2/DE_Analysis_Objs/dds_sm.rds")

###Load these datasets
# STAT1-target genes from Bulk analysis unique(as.data.frame(st_diff_bed_anno)$SYMBOL)
STAT1_diff = readRDS(file = "../ord_et_al_GSE205930/DE_Cell_Specific/STAT1_diff.rds")
#Converting
STAT1_diff_human = gorth(STAT1_diff, source_organism = "mmusculus", target_organism ="hsapiens" )$ortholog_name

STAT1_diff_human = na.omit(STAT1_diff_human)
STAT1_diff_human = STAT1_diff_human[!duplicated(STAT1_diff_human)]

overlap_FC_genes_rna = readRDS(file = "/nvme/mahdi/Analysis/single_cell/atherosclerosis_scRNA/overlap_FC_genes_rna_342.rds")
#Converting
overlap_FC_genes_rna_human = gorth(overlap_FC_genes_rna$gene, source_organism = "mmusculus", target_organism ="hsapiens" )$ortholog_name

overlap_FC_genes_rna_human = na.omit(overlap_FC_genes_rna_human)
overlap_FC_genes_rna_human = overlap_FC_genes_rna_human[!duplicated(overlap_FC_genes_rna_human)]

```

### Lesion vs Non-lesion

```{r echo=FALSE, cache=TRUE}
# SM
dds_hu_sm = readRDS(file = "pan_wirka_hu2/DE_Analysis_Objs/dds_sm.rds")


##########LD vs Ct
res_sm = DESeq2::results(dds_hu_sm, contrast = c("sample_disease_status",
                                        "lesion", "non_lesion"))

res_sm = as.data.frame(res_sm)
res_sm$gene = rownames(res_sm)

# Set threshold for filtering (FDR < 0.05 and log2FC > 0.25)
res_sm_filtd = res_sm %>%
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) >= 0.25 & baseMean > 0.3 ) %>%
  arrange(desc(log2FoldChange))

dim(res_sm_filtd)

#Volcano
res_sm_filtd %>%
  dplyr::filter(abs(log2FoldChange) < 5) %>%
  ggplot(aes(x=log2FoldChange, y=-log10(padj), label= gene)) + 
  geom_point(size = 0.7) + 
  geom_vline(xintercept = c(-0.25, 0, 0.25), color="darkred", linetype="dashed") + 
  geom_hline(yintercept = -log10(1e-10), color="darkred", linetype="dashed") + 
  #geom_text_repel(data = . %>% filter(gene %in% STAT1_diff_human)) +
  theme_classic()

ggsave(filename = "ord_et_al_GSE205930/Figures/CellNo_correction/MQ_NCM_res_LD.png", 
        width = 8.5, height = 9)

ids_sm_filtd = data.frame(symbol = res_sm_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene))
ids_sm_filtd$ENTREZID = mapIds(org.Hs.eg.db, keys = res_sm_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene), keytype = "SYMBOL", column = "ENTREZID", multiVals = "first")
sm_filtd_kegg = enrichKEGG(ids_sm_filtd$ENTREZID, organism = "hsa")

par(mfrow = c(1, 1), mar = c(6, 20, 2, 1))
barplot(height = -log10(sm_filtd_kegg@result$p.adjust)[20:1], names.arg = sm_filtd_kegg@result$Description[20:1],
    horiz = TRUE, las = 1, border = FALSE, cex.names = 0.65, xlab = "-log10(P.adjust)", main = "KEGG Analysis")
abline(v = c(-log10(0.05)), lty = 2)
abline(v = 0, lty = 1)


sm_subset = readRDS(file = "ord_et_al_GSE205930/DE_Cell_Specific/mq_subset_normalized.rds")
sm_subset = ScaleData(sm_subset, assay = "RNA")
DoHeatmap(sm_subset, features = res_sm_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene),  assay = "RNA", size = 4, angle = 90,  group.by = "sample_disease_status") 

DotPlot(sm_subset, features = res_sm_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene),  assay = "RNA", group.by = "sample") + RotatedAxis() + ggtitle("The expression pattern of top 10 genes from pairwaise comparisons") + theme(plot.title = element_text(face = "bold", size = 14))



```

#### Heatmap
```{r}
library(Seurat)
library(tibble)
library(rlang)
library(circlize)
library(ggplot2)
library(ComplexHeatmap)
library(RColorBrewer)
library(grid)
library(grDevices)

hm_limit = c(-2, 0, 2) 
hm_colors = c("#4575b4","white","#d73027")
row_font_size = 6

### Select Genes
#genes = STAT1_diff_human
#genes = overlap_FC_genes_rna_human
#genes = res_sm_filtd %>% arrange(desc(log2FoldChange)) %>% filter(gene %in% STAT1_diff_human) %>% pull(gene)
#genes = res_sm_filtd %>% arrange(desc(log2FoldChange)) %>% filter(gene %in% overlap_FC_genes_rna_human) %>% pull(gene)

genes = res_sm_filtd %>% arrange(desc(log2FoldChange)) %>% pull(gene) # Use this to make the matrix

highlightd_genes = up_list_integ_human
###
sm_subset = ScaleData(sm_subset, assay = "RNA")
mat <- GetAssayData(object = sm_subset, assay = "RNA", slot = "scale.data")
mat <- mat[match(genes, rownames(mat)),]

anno <- sm_subset@meta.data %>%
    rownames_to_column(var = "barcode") %>%
    arrange(!!!syms(c("sample_disease_status", "annotations")))

mat <- t(mat)
mat <- mat[match(anno$barcode, rownames(mat)),]
mat <- t(mat)

annos <- list()

### sample_disease_status
value <- anno[["sample_disease_status"]]
l <- levels(factor(anno[["sample_disease_status"]]))
col_a = c("green", "darkred")
names(col_a) <- l
col_a <- col_a[!is.na(names(col_a))]

ha <- HeatmapAnnotation(a = anno[["sample_disease_status"]],
                              col = list(a = col_a),border = TRUE,
                              annotation_label = "Condition",
                        annotation_legend_param = list(title = "Condition"))

names(ha) <- "sample_disease_status"
annos[["sample_disease_status"]] <- ha
####
####annotations
value <- anno[["annotations"]]
l <- levels(factor(anno[["annotations"]]))
col_b = set_colors("Set2", length(l))
names(col_b) <- l
col_b <- col_b[!is.na(names(col_b))]


ha1 <- HeatmapAnnotation(b = anno[["annotations"]],
                              col = list(b = col_b), border = TRUE,
                              annotation_label = "Cell Type",
                         annotation_legend_param = list(title = "Cell Type"))

names(ha1) <- "annotations"
annos[["annotations"]] <- ha1
###
### arterial_origin
value <- anno[["arterial_origin"]]
l <- levels(factor(anno[["arterial_origin"]]))
col_c = set_colors("Reds", length(l))
names(col_c) <- l
col_c <- col_c[!is.na(names(col_c))]

ha2 <- HeatmapAnnotation(c = anno[["arterial_origin"]],
                              col = list(c = col_c), border = TRUE,
                              annotation_label = "Arterial Bed",
                          annotation_legend_param = list(title = "Arterial Bed"))

names(ha2) <- "arterial_origin"
annos[["arterial_origin"]] <- ha2
####

annos <- do.call(c, annos)
annos@gap <- rep(unit(1,"mm"), length(annos))

ranot = rowAnnotation(link = anno_mark(at = which(rownames(mat) %in% highlightd_genes), 
                     labels = highlightd_genes, side = "left", 
                     labels_gp = gpar(fontsize = 6), padding = unit(1, "mm"))) 
####First approach
set.seed(20) # To be reproducible

ht = Heatmap(mat, cluster_rows = TRUE,
                cluster_columns = FALSE,
                heatmap_legend_param = list(direction = "horizontal",
                                            legend_width = unit(6, "cm"),
                                            title = "Expression"),
                col = colorRamp2(hm_limit, hm_colors),
                show_column_names = FALSE, row_names_side = "left",
                row_names_gp = gpar(fontsize = row_font_size),
                top_annotation = annos, show_row_names = F, show_row_dend = T, row_km = 3,
                row_km_repeats = 20, left_annotation = ranot)
################################
####2nd approach
library(dendsort)

set.seed(20)
row_dend = dendsort(hclust(dist(mat),method = "ward.D")) #"ward.D" looks also OK!
plot(row_dend)

ht = Heatmap(mat, cluster_rows = row_dend, split = 3,
                cluster_columns = FALSE,
                heatmap_legend_param = list(direction = "horizontal",
                                            legend_width = unit(6, "cm"),
                                            title = "Expression"),
                col = colorRamp2(hm_limit, hm_colors),
                show_column_names = FALSE, row_names_side = "left",
                row_names_gp = gpar(fontsize = row_font_size),
                top_annotation = annos, show_row_names = F, show_row_dend = T,
                 left_annotation = ranot)
########       

pdf(file = "pan_wirka_hu2/Figures/htmap_sm_NA.pdf", width = 7.5, height = 9) #using 1st strategy

ht = draw(ht, heatmap_legend_side = "bottom",annotation_legend_side = "right")

dev.off()

saveRDS(ht, file = "pan_wirka_hu2/Figures/htmap_sm_rk3_integ.rds")

############
#open pdf
ht = readRDS(file = "pan_wirka_hu2/Figures/htmap_sm_rk3.rds")
ht
#close pdf
#############

```

#### Venn
```{r}
# From: https://r-graph-gallery.com/14-venn-diagramm
#       https://stackoverflow.com/questions/14243609/venn-diagram-does-not-produce-any-output
library(VennDiagram)

r.dend <- row_dend(ht)  #If needed, extract row dendrogram
rcl.list <- row_order(ht)

# Prepare data sets
cl1 = rownames(mat[rcl.list[["1"]],])
cl1_st = cl1[cl1 %in% up_list_integ_human]

cl2 = rownames(mat[rcl.list[["2"]],])
cl2_st = cl2[cl2 %in% up_list_integ_human]

cl3 = rownames(mat[rcl.list[["3"]],])
cl3_st = cl3[cl3 %in% up_list_integ_human]


# Prepare a palette of 3 colors with R colorbrewer:
library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")[c(1,2)]

# Chart
ven_cl1 = venn.diagram(x = list(cl1, up_list_integ_human),
        category.names = c("Genes in Cluster 1" , "Integrative Genes"),
        filename = NULL, disable.logging = T, output =T,
        # Output features
        imagetype="png" , height = 300, width = 300, resolution = 300, compression = "lzw",
        # Circles
        lwd = 2, lty = 'blank', fill = myCol,
        # Numbers (fontface = "bold")
        cex = 0.9, fontfamily = "sans",
        # Set names (cat.fontface = "bold")
        cat.cex = .7, cat.default.pos = "outer", cat.pos = c(-170, 27),
        cat.fontfamily = "sans")

ven_cl2 = venn.diagram(x = list(cl2, up_list_integ_human),
        category.names = c("Genes in Cluster 2" , "Integrative Genes"),
        filename = NULL, disable.logging = T, output =T,
        # Output features
        imagetype="png" , height = 300, width = 300, resolution = 300, compression = "lzw",
        # Circles
        lwd = 2, lty = 'blank', fill = myCol,
        # Numbers (fontface = "bold")
        cex = 0.9, fontfamily = "sans",
        # Set names (cat.fontface = "bold")
        cat.cex = .7, cat.default.pos = "outer", cat.pos = c(-170, 27),
        cat.fontfamily = "sans")

ven_cl3 = venn.diagram(x = list(cl3, up_list_integ_human),
        category.names = c("Genes in Cluster 3" , "Integrative Genes"),
        filename = NULL, disable.logging = T, output =T,
        # Output features
        imagetype="png" , height = 300, width = 300, resolution = 300, compression = "lzw",
        # Circles
        lwd = 2, lty = 'blank', fill = myCol,
        # Numbers
        cex = 0.9, fontfamily = "sans",
        # Set names
        cat.cex = .7, cat.default.pos = "outer", cat.pos = c(-170, 27),
        cat.fontfamily = "sans")

vens = plot_grid(ven_cl1, ven_cl2, ven_cl3 , nrow = 1, scale =  0.80, align = "hv")

pdf(file = "pan_wirka_hu2/Figures/Ven_sm_rk3.pdf", width = 6, height = 2.5) 

grid.draw(vens)

dev.off()
```

#### ClusterProfiler
```{r}
# cl1
cl1_ranks_df = res_sm_filtd %>% filter(gene %in% cl1) %>% dplyr::select(c(gene, log2FoldChange))
cl1_ranks = cl1_ranks_df$log2FoldChange
names(cl1_ranks) = cl1_ranks_df$gene

# cl2
cl2_ranks_df = res_sm_filtd %>% filter(gene %in% cl2) %>% dplyr::select(c(gene, log2FoldChange))
cl2_ranks = cl2_ranks_df$log2FoldChange
names(cl2_ranks) = cl2_ranks_df$gene

# cl3
cl3_ranks_df = res_sm_filtd %>% filter(gene %in% cl3) %>% dplyr::select(c(gene, log2FoldChange))
cl3_ranks = cl3_ranks_df$log2FoldChange
names(cl3_ranks) = cl3_ranks_df$gene

# Run gene set enrichment
sm_gost_res = gprofiler2::gost(cl3_st, organism = "hsapiens",
                                         ordered_query = T, correction_method = "fdr")
sm_gprofiler_res = sm_gost_res$result
sm_go_bp = sm_gprofiler_res[sm_gprofiler_res$source == "GO:BP", ]

# Plot GO BP terms
sm_go_plot = gprofiler_bar_plot(sm_go_bp, 15, 500, "GO:BP", 15) + new_scale3_bars


ids_mq_filtd = data.frame(symbol = names(cl3_ranks))
ids_mq_filtd$ENTREZID = mapIds(org.Hs.eg.db, keys = names(cl3_ranks), keytype = "SYMBOL", column = "ENTREZID", multiVals = "first")
#mq_filtd_kegg = enrichGO(ids_mq_filtd$ENTREZID, OrgDb = "org.Hs.eg.db", ont = "BP")
mq_filtd_kegg = enrichKEGG(ids_mq_filtd$ENTREZID, organism = "hsa")
par(mfrow = c(1, 1), mar = c(6, 20, 2, 1))
barplot(height = -log10(mq_filtd_kegg@result$p.adjust)[20:1], names.arg = mq_filtd_kegg@result$Description[20:1],
    horiz = TRUE, las = 1, border = FALSE, cex.names = 0.65, xlab = "-log10(P.adjust)", main = "KEGG Analysis")
abline(v = c(-log10(0.05)), lty = 2)
abline(v = 0, lty = 1)

#### ClusterProfiler
# Conversion
cluster_list <- list(Cluster1 = cl1, Cluster2 = cl2, Cluster3 = cl3)
#cluster_list <- list(cl1, cl2, cl2_st, cl3, cl3_st)

map(cluster_list, ~length(.x))

# Function to convert SYMBOL to ENTREZID
convert_to_entrez <- function(gene_cluster) {
  mapIds(org.Hs.eg.db, keys = gene_cluster, keytype = "SYMBOL", column = "ENTREZID", multiVals = "first")
}

# Apply the function to each gene cluster in the list
cluster_list_entrez <- map(cluster_list, convert_to_entrez)

#KEGG
htmp_cluster_list_enrichKEGG = compareCluster(geneCluster = cluster_list_entrez, fun = "enrichKEGG", pvalueCutoff = 0.05, organism = "hsa")
#BP
htmp_cluster_list_enrichGO = compareCluster(geneCluster = cluster_list_entrez, fun = "enrichGO", pvalueCutoff = 0.05, ont="BP", OrgDb = "org.Hs.eg.db")

mq_cls_kegg = dotplot(htmp_cluster_list_enrichKEGG, title = "Comparing KEGG results of heatmap clusters", showCategory=10, font.size =11) + scale_x_discrete(guide = guide_axis(angle = 0)) + theme(axis.title.x = element_blank(), axis.text.x = element_text(size = 10))

pdf(file = "pan_wirka_hu2/Figures/KEGG_Ven_sm_rk3.pdf", width = 8.5, height = 8.2) 

mq_cls_kegg / vens + plot_layout(heights = c(2.7,1))

dev.off()

```


